File: android/app/src/test/java/com/circuvent/hrms/domain/GeofenceTest.kt
package com.circuvent.hrms.domain

import com.circuvent.hrms.domain.Geofence.Confidence
import com.circuvent.hrms.domain.Geofence.Coordinates
import com.circuvent.hrms.domain.Geofence.Fence
import com.circuvent.hrms.domain.Geofence.Severity
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class GeofenceTest {

    private val office = Fence("f1", "Whitefield office", 12.9698, 77.7500, 50.0)

    private fun at(lat: Double, lon: Double, accuracy: Double? = null) =
        Coordinates(latitude = lat, longitude = lon, accuracyMetres = accuracy)

    // ─── The radius ──────────────────────────────────────────

    @Test
    fun `the Earth radius is the one the server uses`() {
        // Load-bearing. The previous generation of this product had the server
        // on 6,371,000 and the phone on 6,371,008.8, so they disagreed by about
        // a metre per kilometre — enough to put somebody on opposite sides of a
        // 50 m office fence depending on which was asked.
        assertEquals(6_371_008.8, Geofence.EARTH_RADIUS_METRES, 0.0)
    }

    @Test
    fun `one degree of latitude is the arc the radius implies`() {
        // This is the radius test with the arithmetic attached: 2*pi*R/360.
        // If somebody "rounds" the constant, this fails with the metres of
        // error rather than with a constant nobody can eyeball.
        val expected = 2 * Math.PI * Geofence.EARTH_RADIUS_METRES / 360.0
        assertEquals(expected, Geofence.distanceMetres(at(0.0, 0.0), at(1.0, 0.0)), 0.001)

        // 111,195.08 m. The neighbouring value 111,194.93 is what the *old*
        // server radius of 6,371,000 produced — the one that disagreed with
        // the phone by about a metre per kilometre. Writing both down is the
        // point: 15 cm per degree is exactly the size of error that reads as a
        // rounding difference and decides a 50 m fence.
        assertEquals(111_195.08, Geofence.distanceMetres(at(0.0, 0.0), at(1.0, 0.0)), 0.05)

        val withOldRadius = 2 * Math.PI * 6_371_000.0 / 360.0
        assertEquals(111_194.93, withOldRadius, 0.05)
    }

    @Test
    fun `a degree of longitude shrinks with latitude`() {
        val equator = Geofence.distanceMetres(at(0.0, 0.0), at(0.0, 1.0))
        val bengaluru = Geofence.distanceMetres(at(12.97, 77.0), at(12.97, 78.0))
        assertTrue(bengaluru < equator)
        // cos(12.97 degrees) is about 0.9745.
        assertEquals(equator * 0.9745, bengaluru, 50.0)
    }

    @Test
    fun `distance to itself is zero and the order does not matter`() {
        assertEquals(0.0, Geofence.distanceMetres(at(12.97, 77.75), at(12.97, 77.75)), 0.0001)
        val a = Geofence.distanceMetres(at(12.97, 77.75), at(12.98, 77.76))
        val b = Geofence.distanceMetres(at(12.98, 77.76), at(12.97, 77.75))
        assertEquals(a, b, 0.0001)
    }

    @Test
    fun `antipodal points do not produce NaN`() {
        // asin of a value that creeps just over 1 through floating point is
        // NaN, and a NaN distance compares false against every fence, which
        // silently reads as "outside".
        val d = Geofence.distanceMetres(at(0.0, 0.0), at(0.0, 180.0))
        assertFalse(d.isNaN())
        assertEquals(Math.PI * Geofence.EARTH_RADIUS_METRES, d, 1.0)
    }

    @Test(expected = IllegalArgumentException::class)
    fun `an impossible latitude is refused rather than measured`() {
        Geofence.distanceMetres(at(91.0, 0.0), at(0.0, 0.0))
    }

    // ─── locateWithin ────────────────────────────────────────

    @Test
    fun `no fences configured is uncertain, not outside`() {
        // "Outside" would refuse the punch. Nobody has said where work is.
        val result = Geofence.locateWithin(at(12.9698, 77.75), emptyList())
        assertEquals(Confidence.UNCERTAIN, result.confidence)
        assertNull(result.fence)
    }

    @Test
    fun `standing at the office is inside`() {
        val result = Geofence.locateWithin(at(12.9698, 77.7500, accuracy = 5.0), listOf(office))
        assertEquals(Confidence.INSIDE, result.confidence)
        assertEquals(office, result.fence)
        assertTrue(result.message.contains("Whitefield"))
    }

    @Test
    fun `a fix too vague to decide a boundary is uncertain, not a refusal`() {
        // A fix accurate to 200 m cannot decide a 50 m fence. Refusing on it
        // would deny a punch on the strength of a reading that says nothing.
        val result = Geofence.locateWithin(at(12.9698, 77.7500, accuracy = 200.0), listOf(office))
        assertEquals(Confidence.UNCERTAIN, result.confidence)
        assertTrue(result.message.contains("accurate to about 200 m"))
    }

    @Test
    fun `well away from the office is outside, with the distance named`() {
        val result = Geofence.locateWithin(at(12.9350, 77.6250, accuracy = 10.0), listOf(office))
        assertEquals(Confidence.OUTSIDE, result.confidence)
        assertNull(result.fence)
        assertTrue("message was: ${result.message}", result.message.contains("km"))
    }

    @Test
    fun `the uncertainty circle must clear the fence to be certainly inside`() {
        // 30 m from the centre of a 50 m fence with a 30 m accuracy: the far
        // edge of the circle is at 60 m, outside. It leans inside and says so.
        val thirtyMetresNorth = at(12.9698 + 0.00027, 77.7500, accuracy = 30.0)
        val result = Geofence.locateWithin(thirtyMetresNorth, listOf(office))
        assertEquals(Confidence.PROBABLY_INSIDE, result.confidence)
        assertEquals(office, result.fence)
    }

    @Test
    fun `the nearest fence is chosen, whatever the order of the list`() {
        val far = Fence("f2", "Electronic City", 12.8400, 77.6800, 50.0)
        val here = at(12.9698, 77.7500, accuracy = 5.0)

        assertEquals(office, Geofence.locateWithin(here, listOf(far, office)).fence)
        assertEquals(office, Geofence.locateWithin(here, listOf(office, far)).fence)
    }

    // ─── Spoofing signals ────────────────────────────────────

    @Test
    fun `a mock provider is the only high severity signal on its own`() {
        val signals = Geofence.spoofSignals(
            Coordinates(12.97, 77.75, accuracyMetres = 8.0, capturedAt = 1_000_000L, isMocked = true),
            now = 1_000_000L,
        )
        assertEquals(1, signals.size)
        assertEquals(Severity.HIGH, signals.first().severity)
        assertEquals("mock_provider", signals.first().code)
    }

    @Test
    fun `sub-metre accuracy is implausible on consumer hardware`() {
        val signals = Geofence.spoofSignals(
            Coordinates(12.97, 77.75, accuracyMetres = 0.4, capturedAt = 1_000L),
            now = 1_000L,
        )
        assertTrue(signals.any { it.code == "implausible_accuracy" })
    }

    @Test
    fun `a timestamp of zero is still a timestamp`() {
        // The defect this pins: the previous implementation tested the
        // timestamp for truthiness, and 0 is falsy, so the staleness and
        // impossible-speed checks were skipped entirely for it. Same shape as
        // treating an empty string as an absent number.
        val signals = Geofence.spoofSignals(
            Coordinates(12.97, 77.75, accuracyMetres = 8.0, capturedAt = 0L),
            now = 60 * 60 * 1000L,
        )
        assertTrue("A fix an hour old should be stale", signals.any { it.code == "stale_fix" })
    }

    @Test
    fun `a plausible commute raises nothing`() {
        val previous = Coordinates(12.9350, 77.6250, capturedAt = 0L)
        val now = Coordinates(12.9698, 77.7500, accuracyMetres = 10.0, capturedAt = 30 * 60 * 1000L)
        val signals = Geofence.spoofSignals(now, previous, now = 30 * 60 * 1000L)
        assertTrue("Signals were: $signals", signals.none { it.code == "impossible_speed" })
    }

    @Test
    fun `crossing the country in a minute is flagged`() {
        val previous = Coordinates(12.97, 77.75, capturedAt = 0L)
        val now = Coordinates(28.61, 77.21, accuracyMetres = 10.0, capturedAt = 60_000L)
        val signals = Geofence.spoofSignals(now, previous, now = 60_000L)
        val speed = signals.firstOrNull { it.code == "impossible_speed" }
        assertTrue("Signals were: $signals", speed != null)
        // Medium, not high. A passenger on a plane produces this too, and an
        // automatic refusal on it would deny somebody a day's pay.
        assertEquals(Severity.MEDIUM, speed!!.severity)
    }

    // ─── The verdict ─────────────────────────────────────────

    @Test
    fun `outside the fence refuses the punch and does not ask for review`() {
        val verdict = Geofence.evaluateClockIn(
            at(12.9350, 77.6250, accuracy = 10.0),
            listOf(office),
        )
        assertFalse(verdict.allowed)
        assertFalse("A refusal needs no review; it never happened", verdict.requiresReview)
    }

    @Test
    fun `inside the fence with a clean fix is allowed and unreviewed`() {
        val verdict = Geofence.evaluateClockIn(
            Coordinates(12.9698, 77.7500, accuracyMetres = 5.0, capturedAt = 1_000L),
            listOf(office),
            now = 1_000L,
        )
        assertTrue(verdict.allowed)
        assertFalse(verdict.requiresReview)
    }

    @Test
    fun `a mocked position is allowed but flagged for a human`() {
        // Not refused. The punch is recorded and a person is asked to look —
        // refusing on a heuristic is how somebody loses a day's pay to a
        // developer setting they forgot was on.
        val verdict = Geofence.evaluateClockIn(
            Coordinates(12.9698, 77.7500, accuracyMetres = 5.0, capturedAt = 1_000L, isMocked = true),
            listOf(office),
            now = 1_000L,
        )
        assertTrue(verdict.allowed)
        assertTrue(verdict.requiresReview)
    }

    @Test
    fun `field staff clock in from anywhere, and are still watched`() {
        val verdict = Geofence.evaluateClockIn(
            Coordinates(0.0, 0.0, accuracyMetres = 5.0, capturedAt = 1_000L, isMocked = true),
            emptyList(),
            now = 1_000L,
            allowAnywhere = true,
        )
        assertTrue(verdict.allowed)
        assertTrue("A mock provider still deserves a look", verdict.requiresReview)
    }

    @Test
    fun `an uncertain fix is allowed but reviewed`() {
        // The punch is not lost because the GPS was poor, and the record says
        // the location could not be confirmed.
        val verdict = Geofence.evaluateClockIn(
            Coordinates(12.9698, 77.7500, accuracyMetres = 200.0, capturedAt = 1_000L),
            listOf(office),
            now = 1_000L,
        )
        assertTrue(verdict.allowed)
        assertTrue(verdict.requiresReview)
        assertEquals(Confidence.UNCERTAIN, verdict.confidence)
    }
}

File: src/lib/mobile/geofence.test.ts
// @vitest-environment node
//
// Clock-in location decides pay, and there are published apps whose only
// purpose is to feed a fake location to attendance software. These tests pin
// both halves of that: the spoofing signals that should be raised, and the
// innocent cases that must not cost someone a day's wages.

import { describe, expect, it } from "vitest";
import {
  distanceMetres,
  evaluateClockIn,
  locateWithin,
  spoofSignals,
  type Coordinates,
  type Geofence,
} from "@/lib/mobile/geofence";

/** Circuvent's Bangalore office, near Koramangala. */
const office: Geofence = {
  id: "blr",
  name: "Bangalore office",
  latitude: 12.9352,
  longitude: 77.6245,
  radiusMetres: 100,
};

const chennai: Geofence = {
  id: "maa",
  name: "Chennai office",
  latitude: 13.0827,
  longitude: 80.2707,
  radiusMetres: 100,
};

function at(over: Partial<Coordinates> = {}): Coordinates {
  return { latitude: office.latitude, longitude: office.longitude, ...over };
}

describe("distanceMetres", () => {
  it("is zero for the same point", () => {
    expect(distanceMetres(at(), at())).toBeCloseTo(0, 5);
  });

  it("measures a short distance accurately", () => {
    // 0.001 degrees of latitude is about 111 m anywhere on Earth.
    const north = at({ latitude: office.latitude + 0.001 });
    expect(distanceMetres(at(), north)).toBeGreaterThan(105);
    expect(distanceMetres(at(), north)).toBeLessThan(115);
  });

  it("measures a long distance accurately", () => {
    // Bangalore to Chennai is roughly 290 km.
    const km = distanceMetres(at(), chennai) / 1000;
    expect(km).toBeGreaterThan(280);
    expect(km).toBeLessThan(300);
  });

  it("is symmetric", () => {
    expect(distanceMetres(at(), chennai)).toBeCloseTo(distanceMetres(chennai, at()), 5);
  });

  it("handles the antimeridian without going the long way round", () => {
    const west = { latitude: 0, longitude: 179.9 };
    const east = { latitude: 0, longitude: -179.9 };
    // About 22 km apart, not most of the way round the planet.
    expect(distanceMetres(west, east) / 1000).toBeLessThan(30);
  });

  it("rejects impossible coordinates rather than returning NaN", () => {
    expect(() => distanceMetres(at({ latitude: 95 }), at())).toThrow(/between -90 and 90/);
    expect(() => distanceMetres(at({ longitude: 200 }), at())).toThrow(/between -180 and 180/);
    expect(() => distanceMetres(at({ latitude: NaN }), at())).toThrow(/finite/);
  });
});

describe("locateWithin", () => {
  it("places a precise fix inside the fence", () => {
    const result = locateWithin(at({ accuracyMetres: 5 }), [office]);
    expect(result.confidence).toBe("inside");
    expect(result.fence?.id).toBe("blr");
  });

  it("places a distant fix outside", () => {
    const result = locateWithin({ ...chennai, accuracyMetres: 5 }, [office]);
    expect(result.confidence).toBe("outside");
    expect(result.fence).toBeNull();
  });

  it("treats accuracy as a radius, not a decoration", () => {
    // Dead centre, but a fix accurate only to 500 m says nothing about a 100 m
    // fence.
    const result = locateWithin(at({ accuracyMetres: 500 }), [office]);
    expect(result.confidence).toBe("uncertain");
    expect(result.message).toMatch(/accurate to about 500 m/);
  });

  it("reports uncertainty when the error circle straddles the boundary", () => {
    // Indoors and near tall buildings this is the common case, and collapsing
    // it either way marks an office worker absent for standing by a window.
    const nearEdge = at({ latitude: office.latitude + 0.0009, accuracyMetres: 60 });
    const result = locateWithin(nearEdge, [office]);
    expect(["probably_inside", "uncertain"]).toContain(result.confidence);
  });

  it("requires the whole error circle inside before saying so with certainty", () => {
    const nearEdge = at({ latitude: office.latitude + 0.0007, accuracyMetres: 40 });
    const result = locateWithin(nearEdge, [office]);
    expect(result.confidence).not.toBe("inside");
  });

  it("picks the nearest of several fences", () => {
    const result = locateWithin(at({ accuracyMetres: 5 }), [chennai, office]);
    expect(result.fence?.id).toBe("blr");
  });

  it("says so when no fences are configured", () => {
    const result = locateWithin(at(), []);
    expect(result.confidence).toBe("uncertain");
    expect(result.message).toMatch(/No work locations/);
  });

  it("reports how far outside, in readable units", () => {
    const result = locateWithin({ ...chennai, accuracyMetres: 5 }, [office]);
    expect(result.message).toMatch(/km from Bangalore office/);
  });

  it("honours a custom accuracy ceiling", () => {
    const result = locateWithin(at({ accuracyMetres: 150 }), [office], {
      maxAccuracyMetres: 100,
    });
    expect(result.confidence).toBe("uncertain");
  });
});

describe("spoofSignals", () => {
  it("raises nothing for an ordinary fix", () => {
    expect(spoofSignals(at({ accuracyMetres: 12, capturedAt: 1_000 }), undefined, 1_000)).toEqual(
      []
    );
  });

  it("raises a high-severity signal for a mock provider", () => {
    // The only signal the operating system asserts rather than something we
    // inferred.
    const signals = spoofSignals(at({ isMocked: true }));
    expect(signals[0]).toMatchObject({ code: "mock_provider", severity: "high" });
  });

  it("flags implausibly perfect accuracy", () => {
    // Consumer GPS does not achieve sub-3 m; a faker often picks a number that
    // sounded good.
    const signals = spoofSignals(at({ accuracyMetres: 1 }));
    expect(signals.map((s) => s.code)).toContain("implausible_accuracy");
  });

  it("does not flag ordinary good accuracy", () => {
    expect(spoofSignals(at({ accuracyMetres: 5 })).map((s) => s.code)).not.toContain(
      "implausible_accuracy"
    );
  });

  it("flags a stale fix", () => {
    const signals = spoofSignals(at({ capturedAt: 0 }), undefined, 10 * 60_000);
    expect(signals.map((s) => s.code)).toContain("stale_fix");
  });

  it("does not flag a fresh fix", () => {
    const signals = spoofSignals(at({ capturedAt: 0 }), undefined, 60_000);
    expect(signals.map((s) => s.code)).not.toContain("stale_fix");
  });

  it("flags movement no vehicle could make", () => {
    // Bangalore to Chennai in one minute.
    const previous = at({ capturedAt: 0 });
    const now = { ...chennai, capturedAt: 60_000 };
    const signals = spoofSignals(now, previous, 60_000);

    expect(signals.map((s) => s.code)).toContain("impossible_speed");
  });

  it("checks speed even when the earlier timestamp is zero", () => {
    // A timestamp of 0 is a real value. Guarding with `if (previous.capturedAt)`
    // skips the whole check for it — the same shape of bug as treating
    // Number("") as absent, and the case above happens to exercise it.
    const previous = at({ capturedAt: 0 });
    const jump = { ...chennai, capturedAt: 1_000 };

    expect(spoofSignals(jump, previous, 1_000).map((s) => s.code)).toContain(
      "impossible_speed"
    );
  });

  it("does not flag air travel, because people clock in from airports", () => {
    // 290 km in an hour is about 290 km/h — fast, but a train manages it and a
    // plane far exceeds it. The threshold sits above airliner cruise.
    const previous = at({ capturedAt: 0 });
    const now = { ...chennai, capturedAt: 3_600_000 };

    expect(spoofSignals(now, previous, 3_600_000).map((s) => s.code)).not.toContain(
      "impossible_speed"
    );
  });

  it("does not divide by zero on two fixes at the same instant", () => {
    const previous = at({ capturedAt: 1_000 });
    expect(() => spoofSignals({ ...chennai, capturedAt: 1_000 }, previous)).not.toThrow();
  });

  it("raises several signals at once", () => {
    const signals = spoofSignals(at({ isMocked: true, accuracyMetres: 1, capturedAt: 0 }), undefined, 600_000);
    expect(signals.length).toBeGreaterThanOrEqual(3);
  });
});

describe("evaluateClockIn", () => {
  it("accepts someone at the office", () => {
    const verdict = evaluateClockIn(at({ accuracyMetres: 8 }), [office]);
    expect(verdict.allowed).toBe(true);
    expect(verdict.requiresReview).toBe(false);
  });

  it("refuses someone plainly elsewhere", () => {
    const verdict = evaluateClockIn({ ...chennai, accuracyMetres: 8 }, [office]);
    expect(verdict.allowed).toBe(false);
  });

  it("accepts but flags a mock provider rather than refusing", () => {
    // Someone whose phone reports a mock provider because a developer tool is
    // installed still turned up for work. Refusing costs a day's pay on a
    // guess; flagging puts it in front of a person who can ask.
    const verdict = evaluateClockIn(at({ accuracyMetres: 8, isMocked: true }), [office]);
    expect(verdict.allowed).toBe(true);
    expect(verdict.requiresReview).toBe(true);
    expect(verdict.signals.map((s) => s.code)).toContain("mock_provider");
  });

  it("accepts but flags an uncertain fix", () => {
    // Indoor GPS is genuinely poor, and the alternative is marking everyone in
    // the middle of a building absent.
    const verdict = evaluateClockIn(at({ accuracyMetres: 400 }), [office]);
    expect(verdict.allowed).toBe(true);
    expect(verdict.requiresReview).toBe(true);
    expect(verdict.confidence).toBe("uncertain");
  });

  it("allows field staff anywhere", () => {
    const verdict = evaluateClockIn({ ...chennai, accuracyMetres: 8 }, [office], {
      allowAnywhere: true,
    });
    expect(verdict.allowed).toBe(true);
    expect(verdict.requiresReview).toBe(false);
  });

  it("still flags a mock provider for field staff", () => {
    const verdict = evaluateClockIn({ ...chennai, isMocked: true }, [office], {
      allowAnywhere: true,
    });
    expect(verdict.allowed).toBe(true);
    expect(verdict.requiresReview).toBe(true);
  });

  it("refuses on distance before considering spoofing", () => {
    // Somebody plainly in another city does not need a review; they need a no.
    const verdict = evaluateClockIn(
      { ...chennai, accuracyMetres: 8, isMocked: true },
      [office]
    );
    expect(verdict.allowed).toBe(false);
    expect(verdict.requiresReview).toBe(false);
  });

  it("returns a message written for the person, not the log", () => {
    const verdict = evaluateClockIn({ ...chennai, accuracyMetres: 8 }, [office]);
    expect(verdict.message).toMatch(/from Bangalore office/);
  });
});
